sdk/java/pulumi/src/main/java/com/pulumi/deployment/internal/DeploymentImpl.java

```diff
@@ -1706,21 +1706,25 @@ public <T> CompletableFuture<Result<T>> runAsync(Supplier<T> callback) {
             registerTask("DefaultRunner#runAsync", valueFuture);
             // loop starts after the callback
             return valueFuture
-                    .thenCompose(__ -> whileRunningAsync())
-                    .handle((code, throwable) -> {
+                    .thenCompose(value -> whileRunningAsync().thenApply(__ -> value))
```

Comment:
Why not this:

```
whileRunningAsync().thenCompose(_ -> valueFuture)
```

Or use `thenCombine` API.

What I am trying to see is that when `valueFuture` takes a long time, but there's an exception out of band, this exception short-circuits `valueFuture` instead of waiting on it. Let me try to make a test.

Reply:
I'm slightly inclined to leave as-is, because this looks more confusing in my opinion:
`valueFuture.thenCompose(__ -> whileRunningAsync().thenCompose(___ -> valueFuture))` 
than this
`valueFuture.thenCompose(value -> whileRunningAsync().thenApply(__ -> value))`

Unless I'm missing something, they are equivalent in behaviour, right? (Tests are green in both cases)